데이터셋 정보

이 데이터 세트는 오듀본 협회 북미 버섯 야외 안내서(1981)에서 발췌한 양송이과(Agaricus)와 덩굴버섯과(Lepiota)에 속하는 23종의 아가미버섯에 대한 가상 표본 설명을 포함합니다. 각 종은 확실히 식용 가능, 확실히 독성, 또는 식용 가능 여부가 불확실하고 권장하지 않는 것으로 분류됩니다. 마지막 분류는 독성 분류와 통합되었습니다. 안내서에는 버섯의 식용 여부를 판단하는 간단한 규칙이 없으며, 옻나무나 담쟁이덩굴처럼 "잎이 세 개면 먹지 마라"와 같은 규칙은 없다고 명시되어 있습니다.

### 칼럼 정리

모든 칼럼은 알파벳 한 글자 코드로 된 범주형 데이터입니다. `class`가 예측할 타깃(정답)이고, 나머지 22개는 특성입니다. 값은 실제 CSV에 들어 있는 값만 적었습니다.

| # | 칼럼 | 의미 | 값 |
|---|---|---|---|
| 1 | `class` | **분류(타깃)** | e=식용, p=독성 |
| 2 | `cap-shape` | 갓 모양 | b=종형, c=원뿔형, x=볼록형, f=평면형, k=뭉툭한형, s=오목한형 |
| 3 | `cap-surface` | 갓 표면 | f=섬유질, g=홈, y=비늘 모양, s=매끄러움 |
| 4 | `cap-color` | 갓 색상 | n=갈색, b=연한 갈색, c=계피색, g=회색, r=녹색, p=분홍, u=보라, e=빨강, w=흰색, y=노랑 |
| 5 | `bruises` | 멍(상처 났을 때 변색되는지) | t=있음, f=없음 |
| 6 | `odor` | 냄새 | a=아몬드, l=아니스, c=크레오소트, y=생선, f=불쾌한, m=곰팡이, n=없음, p=자극적, s=매운 |
| 7 | `gill-attachment` | 주름(아가미)이 줄기에 붙은 방식 | a=부착형, f=자유형 |
| 8 | `gill-spacing` | 주름 간격 | c=좁음, w=빽빽함 |
| 9 | `gill-size` | 주름 크기 | b=넓음, n=좁음 |
| 10 | `gill-color` | 주름 색상 | k=검정, n=갈색, b=황갈색, h=초콜릿, g=회색, r=녹색, o=주황, p=분홍, u=보라, e=빨강, w=흰색, y=노랑 |
| 11 | `stalk-shape` | 줄기 모양 | e=아래로 갈수록 넓어짐(확대), t=가늘어짐 |
| 12 | `stalk-root` | 줄기 뿌리 모양 | b=구근형, c=곤봉형, e=동일형, r=뿌리내림, **?=누락(결측치)** |
| 13 | `stalk-surface-above-ring` | 고리 위 줄기 표면 | f=섬유질, y=비늘, k=실크, s=매끄러움 |
| 14 | `stalk-surface-below-ring` | 고리 아래 줄기 표면 | 13번과 같음 |
| 15 | `stalk-color-above-ring` | 고리 위 줄기 색상 | n=갈색, b=연한 갈색, c=계피색, g=회색, o=주황, p=분홍, e=빨강, w=흰색, y=노랑 |
| 16 | `stalk-color-below-ring` | 고리 아래 줄기 색상 | 15번과 같음 |
| 17 | `veil-type` | 베일(막) 유형 | p=부분 |
| 18 | `veil-color` | 베일 색상 | n=갈색, o=주황, w=흰색, y=노랑 |
| 19 | `ring-number` | 줄기에 달린 고리 개수 | n=없음, o=1개, t=2개 |
| 20 | `ring-type` | 고리 유형 | e=소멸형, f=확산형, l=대형, n=없음, p=펜던트형 |
| 21 | `spore-print-color` | 포자 무늬 색상 | k=검정, n=갈색, b=황갈색, h=초콜릿, r=녹색, o=주황, u=보라, w=흰색, y=노랑 |
| 22 | `population` | 개체군 분포(어떻게 모여 자라는지) | a=풍부함, c=밀집됨, n=다수, s=분산됨, v=여러 개, y=고립됨 |
| 23 | `habitat` | 서식지 | g=풀밭, l=나뭇잎, m=초원, p=길, u=도시, w=쓰레기장, d=숲 |

In [2]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv('mushrooms.csv')
df.head(5) 


,class,cap-shape,cap-surface,cap-color,bruises,odor,gill-attachment,gill-spacing,gill-size,gill-color,...,stalk-surface-below-ring,stalk-color-above-ring,stalk-color-below-ring,veil-type,veil-color,ring-number,ring-type,spore-print-color,population,habitat
0,p,x,s,n,t,p,f,c,n,k,...,s,w,w,p,w,o,p,k,s,u
1,e,x,s,y,t,a,f,c,b,k,...,s,w,w,p,w,o,p,n,n,g
2,e,b,s,w,t,l,f,c,b,n,...,s,w,w,p,w,o,p,n,n,m
3,p,x,y,w,t,p,f,c,n,n,...,s,w,w,p,w,o,p,k,s,u
4,e,x,s,g,f,n,f,w,b,k,...,s,w,w,p,w,o,e,n,a,g


In [4]:
# 데이터 타입과 결측치 확인
df.describe()
print(df.isna().sum().sum())   # NaN 개수 합계
print((df == '?').sum()[(df == '?').sum() > 0])   # '?'로 숨은 결측

0
stalk-root    2480
dtype: int64


23개 칼럼 모두 object(문자) 타입 → 모델에 넣으려면 인코딩 필요.
NaN은 0개지만, stalk-root에 '?'로 표시된 결측이 2,480개 숨어 있음.

In [5]:
print(df.shape)
print(df['class'].value_counts())
print(df['stalk-root'].value_counts())

(8124, 23)
class
e    4208
p    3916
Name: count, dtype: int64
stalk-root
b    3776
?    2480
e    1120
c     556
r     192
Name: count, dtype: int64


식용(e) 4208 개 / 독성(p) 3916개 - 거의 반반임.

앞서 칼럼 정보에서 'stalk-root' 혼자 ?(결측치)를 가지고 있었으므로 value_counts()로 확인. 총 2480개의 누락 확인.

In [6]:
# stalk-root(줄기 뿌리 모양) 과의 관계 확인하기
#pd.crosstab(df['stalk-root'], df['class'], normalize='index')
df.groupby('stalk-root')['class'].value_counts(normalize=True)

stalk-root  class
?           p        0.709677
            e        0.290323
b           e        0.508475
            p        0.491525
c           e        0.920863
            p        0.079137
e           e        0.771429
            p        0.228571
r           e        1.000000
Name: proportion, dtype: float64

groupby()로 그룹별 class 비율 확인

stalk-root 값(b, c, e, r, ?)을 행으로 class값(e, p)을 열로 놓고
각 조합이 몇 번 나왔는지 알려주는 것이다.

독성(p) 비율이 71%로, 전체 평균보다 높다.
결측 자체가 독성 신호라는 뜻.
결측값을 처리X

In [7]:
# nunique()로 컬럼마다 고유값이 몇 개인지 확인.
print(df.nunique())

class                        2
cap-shape                    6
cap-surface                  4
cap-color                   10
bruises                      2
odor                         9
gill-attachment              2
gill-spacing                 2
gill-size                    2
gill-color                  12
stalk-shape                  2
stalk-root                   5
stalk-surface-above-ring     4
stalk-surface-below-ring     4
stalk-color-above-ring       9
stalk-color-below-ring       9
veil-type                    1
veil-color                   4
ring-number                  3
ring-type                    5
spore-print-color            9
population                   6
habitat                      7
dtype: int64


veil-type(막 유형) - 고유값이 1개뿐이라 모든 행이 같은 값이라는 뜻/
분류에 아무 정보를 주지 못함.

In [8]:
# veil-type 삭제.
df = df.drop('veil-type', axis=1)

In [9]:
# 레이블 인코딩
# 대상 : class (e/p)
from sklearn.preprocessing import LabelEncoder

encoder = LabelEncoder()
# 어떤 값들이 있는지 학습
encoder.fit(df['class']) 
# 학습한 기준대로 변환 (e-> 0, p->1)
df['class'] = encoder.transform(df['class']) 


print(encoder.classes_)



['e' 'p']


In [10]:
# 피처 원-핫 인코딩
# 대상 : 나머지 21개 컬럼(갓 모양, 냄새, 색깔 등)
# 21개 칼럼은 원-핫 인코딩 진행하는 이유는 숫자의 가중치에 따른 문제를
# 해결하기 위함.
# 모든 칼럼의 dtype이 object 였으므로 진행.

X = pd.get_dummies(df.drop('class', axis=1), dtype=int)
y = df['class']

print('인코딩 전:', df.drop('class', axis=1).shape)
print('인코딩 후:', X.shape)

인코딩 전: (8124, 21)
인코딩 후: (8124, 116)


인코딩 후 칼럼 수가 21개 칼럼 각각의 고유값 개수 만큼 늘어남.
21 -> 116

In [11]:
# 전처리 후 새 DataFrame 저장

# 8124개 모두 사용했으나 필요 여부에 따라 샘플링을 할지 말지 결정해야함.

new_data = pd.concat([X, y], axis=1)
new_data.to_csv('mushrooms_preprocessed.csv', index=False)

# 원본 범주명 인코딩 전 파일
df_raw = pd.read_csv('mushrooms.csv').drop('veil-type', axis=1)
df_raw.to_csv('mushrooms_cleaned.csv', index=False)

